NamedEhCacheProvider doesn't allow explicitly configured cache attributes

https://www.playframework.com/documentation/2.4.x/ScalaCache states that additional caches can be configured by creating a custom ehcache.xml.  It also indicates that you should list these caches in play.cache.bindCaches and use the NamedCache qualifier to access these caches.  This doesn't appear to work as documented (tested with Play 2.4.6).

I've defined a custom ehcache.xml:

```
<ehcache xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="http://www.ehcache.org/ehcache.xsd" updateCheck="false">

  <defaultCache
    maxEntriesLocalHeap="10000"
    eternal="false"
    timeToIdleSeconds="120"
    timeToLiveSeconds="120"
    overflowToDisk="false"
    maxElementsOnDisk="10000000"
    diskPersistent="false"
    diskExpiryThreadIntervalSeconds="120"
    memoryStoreEvictionPolicy="LRU" />

  <cache name="my-cache"
    maxEntriesLocalHeap="10000" />

</ehcache>
```

In application.conf, I've listed this cache:

```
play.cache.bindCaches = ["my-cache"]
```

And from my code I've wired the cache up:

```
class Test @Inject()(@NamedCache("my-cache") cache: CacheApi)
```


If I try to start the application with this configuration, it blows up:

```
Caused by: net.sf.ehcache.ObjectExistsException: Cache my-cache already exists
  at net.sf.ehcache.CacheManager.addCache(CacheManager.java:1171)
  at play.api.cache.NamedEhCacheProvider.get$lzycompute(Cache.scala:208)
```

If I remove the custom cache definitions from ehcache.xml, things work, but this means I cannot set custom attributes on individual caches.

At first glance it seems like the code should be changed to something more like this:

```
diff --git a/framework/src/play-cache/src/main/scala/play/api/cache/Cache.scala b/framework/src/play-cache/src/main/scala/play/api/cache/Cache.scala
index d55bb23..970a40a 100644
--- a/framework/src/play-cache/src/main/scala/play/api/cache/Cache.scala
+++ b/framework/src/play-cache/src/main/scala/play/api/cache/Cache.scala
@@ -205,8 +205,10 @@ class CacheManagerProvider @Inject() (env: Environment, config: Configuration, l
 private[play] class NamedEhCacheProvider(name: String) extends Provider[Ehcache] {
   @Inject private var manager: CacheManager = _
   lazy val get: Ehcache = {
-    manager.addCache(name)
-    manager.getEhcache(name)
+    Option(manager.getEhcache(name)).getOrElse {
+      manager.addCache(name)
+      manager.getEhcache(name)
+    }
   }
 }

```

--- a/documentation/manual/working/javaGuide/main/cache/JavaCache.md
+++ b/documentation/manual/working/javaGuide/main/cache/JavaCache.md
@@ -3,7 +3,7 @@
 
 Caching data is a typical optimization in modern applications, and so Play provides a global cache. An important point about the cache is that it behaves just like a cache should: the data you just stored may just go missing.
 
-For any data stored in the cache, a regeneration strategy needs to be put in place in case the data goes missing. This philosophy is one of the fundamentals behind Play, and is different from Java EE, where the session is expected to retain values throughout its lifetime. 
+For any data stored in the cache, a regeneration strategy needs to be put in place in case the data goes missing. This philosophy is one of the fundamentals behind Play, and is different from Java EE, where the session is expected to retain values throughout its lifetime.
 
 The default implementation of the cache API uses [EHCache](http://www.ehcache.org/).
 
@@ -49,13 +49,17 @@ If you want to access multiple different ehcache caches, then you'll need to tel
 
     play.cache.bindCaches = ["db-cache", "user-cache", "session-cache"]
 
+By default, Play will try to create these caches for you. If you would like to define them yourself in `ehcache.xml`, you can set:
+
+    play.cache.createBoundCaches = false
+
 Now to access these different caches, when you inject them, use the [NamedCache](api/java/play/cache/NamedCache.html) qualifier on your dependency, for example:
 
 @[qualified](code/javaguide/cache/qualified/Application.java)
 
 ## Caching HTTP responses
 
-You can easily create a smart cached action using standard `Action` composition. 
+You can easily create a smart cached action using standard `Action` composition.
 
 > **Note:** Play HTTP `Result` instances are safe to cache and reuse later.
 

--- a/documentation/manual/working/scalaGuide/main/cache/ScalaCache.md
+++ b/documentation/manual/working/scalaGuide/main/cache/ScalaCache.md
@@ -5,7 +5,7 @@ Caching data is a typical optimization in modern applications, and so Play provi
 
 > An important point about the cache is that it behaves just like a cache should: the data you just stored may just go missing.
 
-For any data stored in the cache, a regeneration strategy needs to be put in place in case the data goes missing. This philosophy is one of the fundamentals behind Play, and is different from Java EE, where the session is expected to retain values throughout its lifetime. 
+For any data stored in the cache, a regeneration strategy needs to be put in place in case the data goes missing. This philosophy is one of the fundamentals behind Play, and is different from Java EE, where the session is expected to retain values throughout its lifetime.
 
 The default implementation of the Cache API uses [EHCache](http://ehcache.org/).
 
@@ -55,6 +55,10 @@ It is possible to access different caches.  The default cache is called `play`, 
 If you want to access multiple different ehcache caches, then you'll need to tell Play to bind them in `application.conf`, like so:
 
     play.cache.bindCaches = ["db-cache", "user-cache", "session-cache"]
+
+By default, Play will try to create these caches for you. If you would like to define them yourself in `ehcache.xml`, you can set:
+
+    play.cache.createBoundCaches = false
 
 Now to access these different caches, when you inject them, use the [NamedCache](api/java/play/cache/NamedCache.html) qualifier on your dependency, for example:
 

--- a/framework/src/play-cache/src/main/resources/reference.conf
+++ b/framework/src/play-cache/src/main/resources/reference.conf
@@ -9,6 +9,9 @@ play {
     configResource = "ehcache.xml"
     # The caches to bind
     bindCaches = []
+    # If true, play will try to create the caches in bindCaches. If false, the caches in bindCaches must already be
+    # defined in ehcache.xml
+    createBoundCaches = true
     # The name of the default cache to use in ehcache
     defaultCache = "play"
   }

--- a/framework/src/play-cache/src/main/scala/play/api/cache/Cache.scala
+++ b/framework/src/play-cache/src/main/scala/play/api/cache/Cache.scala
@@ -152,7 +152,8 @@ trait EhCacheComponents {
    * Use this to create with the given name.
    */
   def cacheApi(name: String): CacheApi = {
-    new EhCacheApi(NamedEhCacheProvider.getNamedCache(name, ehCacheManager))
+    val createNamedCaches = configuration.underlying.getBoolean("play.cache.createBoundCaches")
+    new EhCacheApi(NamedEhCacheProvider.getNamedCache(name, ehCacheManager, createNamedCaches))
   }
 
   lazy val defaultCacheApi: CacheApi = cacheApi("play")
@@ -169,6 +170,7 @@ class EhCacheModule extends Module {
   def bindings(environment: Environment, configuration: Configuration) = {
     val defaultCacheName = configuration.underlying.getString("play.cache.defaultCache")
     val bindCaches = configuration.underlying.getStringList("play.cache.bindCaches").toSeq
+    val createBoundCaches = configuration.underlying.getBoolean("play.cache.createBoundCaches")
 
     // Creates a named cache qualifier
     def named(name: String): NamedCache = {
@@ -181,7 +183,7 @@ class EhCacheModule extends Module {
       val ehcacheKey = bind[Ehcache].qualifiedWith(namedCache)
       val cacheApiKey = bind[CacheApi].qualifiedWith(namedCache)
       Seq(
-        ehcacheKey.to(new NamedEhCacheProvider(name)),
+        ehcacheKey.to(new NamedEhCacheProvider(name, createBoundCaches)),
         cacheApiKey.to(new NamedCacheApiProvider(ehcacheKey)),
         bind[JavaCacheApi].qualifiedWith(namedCache).to(new NamedJavaCacheApiProvider(cacheApiKey)),
         bind[Cached].qualifiedWith(namedCache).to(new NamedCachedProvider(cacheApiKey))
@@ -208,21 +210,28 @@ class CacheManagerProvider @Inject() (env: Environment, config: Configuration, l
   }
 }
 
-private[play] class NamedEhCacheProvider(name: String) extends Provider[Ehcache] {
+private[play] class NamedEhCacheProvider(name: String, create: Boolean) extends Provider[Ehcache] {
   @Inject private var manager: CacheManager = _
-  lazy val get: Ehcache = NamedEhCacheProvider.getNamedCache(name, manager)
+  lazy val get: Ehcache = NamedEhCacheProvider.getNamedCache(name, manager, create)
 }
 
 private[play] object NamedEhCacheProvider {
-  def getNamedCache(name: String, manager: CacheManager) = try {
-    manager.addCache(name)
+  def getNamedCache(name: String, manager: CacheManager, create: Boolean) = try {
+    if (create) {
+      manager.addCache(name)
+    }
     manager.getEhcache(name)
   } catch {
     case e: ObjectExistsException =>
       throw new EhCacheExistsException(
         s"""An EhCache instance with name '$name' already exists.
            |
-           |This usually indicates that multiple instances of a dependent component (e.g. a Play application) have been started at the same time.
+           |This may indicate that multiple instances of a dependent component (e.g. a Play application) have started at the same time.
+           |
+           |If you would like to configure caches in ehcache.xml instead of having Play create them, please set:
+           |
+           |    play.cache.createBoundCaches = false
+           |
          """.stripMargin, e)
   }
 }

--- a/framework/src/play-cache/src/test/scala/play/api/cache/CacheApiSpec.scala
+++ b/framework/src/play-cache/src/test/scala/play/api/cache/CacheApiSpec.scala
@@ -1,0 +1,43 @@
+package play.api.cache
+
+import javax.inject.{ Inject, Provider }
+
+import net.sf.ehcache.CacheManager
+import play.api.inject._
+import play.api.test.{ PlaySpecification, WithApplication }
+
+class CacheApiSpec extends PlaySpecification {
+  sequential
+
+  "CacheApi" should {
+    "bind named caches" in new WithApplication(
+      _.configure(
+        "play.cache.bindCaches" -> Seq("custom")
+      )
+    ) {
+      app.injector.instanceOf[NamedCacheController]
+    }
+    "bind already created named caches" in new WithApplication(
+      _.overrides(
+        bind[CacheManager].toProvider[CustomCacheManagerProvider]
+      ).configure(
+          "play.cache.createBoundCaches" -> false,
+          "play.cache.bindCaches" -> Seq("custom")
+        )
+    ) {
+      app.injector.instanceOf[NamedCacheController]
+    }
+  }
+}
+
+class CustomCacheManagerProvider @Inject() (cacheManagerProvider: CacheManagerProvider) extends Provider[CacheManager] {
+  lazy val get = {
+    val mgr = cacheManagerProvider.get
+    mgr.removalAll()
+    mgr.addCache("custom")
+    mgr
+  }
+}
+
+class NamedCacheController @Inject() (
+  @NamedCache("custom") val cache: CacheApi)

